In [0]:
import pandas as pd
from plotly.subplots import make_subplots
import pandas as pd
import plotly.graph_objects as go
import seaborn as sns

In [0]:
team_all_data = pd.read_csv('/content/drive/My Drive/Colab Notebooks/Pro-Kabbadi All Seasons Data/All_teams_All_Season_Latest.csv')

In [7]:
team_all_data.head()

Unnamed: 0  match_played  ...  team_total_points team_total_points_rank
0           0            14  ...                405                      8
1           1            14  ...                430                      7
2           2            14  ...                456                      6
3           3            16  ...                550                      2
4           4            16  ...                504                      3

[5 rows x 33 columns]

In [0]:
player_df = pd.read_csv('/content/drive/My Drive/Colab Notebooks/Pro-Kabbadi All Seasons Data/All_Players_All_Seasons_Latest.csv')

In [0]:
player_merged_data = pd.read_csv('/content/drive/My Drive/Colab Notebooks/Pro-Kabbadi All Seasons Data/merge_it.csv')

In [10]:
player_df.head()

player_id  player_team  ... unsuccess_raids  yellow_cards
0        105            0  ...              12             0
1        105            0  ...              25             0
2        105            0  ...              19             0
3        105            0  ...              10             0
4        105            0  ...              21             1

[5 rows x 39 columns]

In [0]:
player_all_data = pd.merge(player_df,player_merged_data, on=['team_name','season'])

In [0]:
df = player_all_data.copy()

In [32]:
player_all_data.describe()

player_id  player_team  ...  succ_raid_percent  succ_tackle_percent
count   650.000000        650.0  ...         650.000000           650.000000
mean    755.924615          0.0  ...          31.529835            26.004690
std     893.943237          0.0  ...          30.845872            20.852218
min      12.000000          0.0  ...           0.000000             0.000000
25%     160.000000          0.0  ...           0.000000             0.000000
50%     318.000000          0.0  ...          33.333333            28.571429
75%     771.250000          0.0  ...          59.045437            42.481061
max    2683.000000          0.0  ...         100.000000           100.000000

[8 rows x 40 columns]

In [0]:
player_all_data['succ_raid_percent']=(player_all_data['success_raids']/(player_all_data['success_raids']+player_all_data['unsuccess_raids']))*100
player_all_data['succ_tackle_percent'] = (player_all_data['success_tackles']/player_all_data['tackles'])*100

In [0]:
team_all_data['spt'] = (team_all_data['team_super_raid'] + team_all_data['team_super_tackles'] + team_all_data['team_all_outs_inflicted']) - team_all_data['team_all_outs_conceded']

In [0]:
player_all_data['succ_raid_percent'].fillna(0,inplace=True)
player_all_data['succ_tackle_percent'].fillna(0,inplace=True)

In [17]:
player_all_data.head()

player_id  player_team  ... succ_raid_percent  succ_tackle_percent
0        105            0  ...         80.952381             7.692308
1        123            0  ...          0.000000            29.230769
2        156            0  ...         57.142857            14.285714
3         54            0  ...          0.000000            51.612903
4         71            0  ...        100.000000            33.707865

[5 rows x 42 columns]

In [0]:

def player_plot_per_season(df,season):
    df1 = df[df.season==season]
    df1 = df1[df1.success_raids > 100]
    df1 = df1.sort_values(by=['succ_raid_percent'],ascending=False).head(10)
    fig = go.Figure(data=[    
    go.Bar(name='succ_raid_percent', x=df1.name, y=df1.succ_raid_percent),
    go.Bar(name='team_total_points_rank', x=df1.name, y=df1.team_total_points_rank)
    ])
    fig.update_layout(showlegend=True,title_text='Season : '+str(season) ,title=dict(x=0.5,y=0.9))
#     fig.update_layout(legend_orientation="h",legend=dict(x=0, y=-0.2))
    fig.update_layout(barmode='group')
    fig.show()

In [0]:
def team_tackle_plot_per_season(df,season):
    df1 = df[df.season==season]
    df1 = df1.sort_values(by=['team_successful_tackles'],ascending=False).head(10)
    fig = go.Figure(data=[    
    go.Bar(name='team_successful_tackles', x=df1.team_name, y=df1.team_successful_tackles),
    go.Bar(name='team_total_points_rank', x=df1.team_name, y=df1.team_total_points_rank)
    ])
    fig.update_layout(showlegend=True,title_text='Season : '+str(season) ,title=dict(x=0.5,y=0.9))
#     fig.update_layout(legend_orientation="h",legend=dict(x=0, y=-0.2))
    fig.update_layout(barmode='group')
    fig.show()

In [0]:
def team_raid_plot_per_season(df,season):
    df1 = df[df.season==season]
    df1 = df1.sort_values(by=['team_successful_raids'],ascending=False).head(10)
    fig = go.Figure(data=[    
    go.Bar(name='team_successful_raids', x=df1.team_name, y=df1.team_successful_raids),
    go.Bar(name='team_total_points_rank', x=df1.team_name, y=df1.team_total_points_rank)
    ])
    fig.update_layout(showlegend=True,title_text='Season : '+str(season) ,title=dict(x=0.5,y=0.9))
#     fig.update_layout(legend_orientation="h",legend=dict(x=0, y=-0.2))
    fig.update_layout(barmode='group')
    fig.show()

In [0]:
def team_player_plot_success_raids(df,season):
    df1 = df[df.season==season]
    #df1 = df1.sort_values(by=['team_successful_raids'],ascending=False)
    fig = go.Figure(data=[    
    #go.Bar(name='player_successful_raids', x=df1.name, y=df1.success_raids),
    go.Bar(name='team_name', x=df1.team_name, y=df1.success_raids,
             text=(df1.name),textposition='auto')
    ])
    fig.update_layout(showlegend=True,title_text='Season : '+str(season) ,title=dict(x=0.5,y=0.9))
#     fig.update_layout(legend_orientation="h",legend=dict(x=0, y=-0.2))
    fig.update_layout(barmode='stack',xaxis={'categoryorder':'array', 'categoryarray':['Dabang Delhi K.C.','Bengal Warriors','Bengaluru Bulls','Haryana Steelers']})
    fig.show()

In [0]:
def team_player_plot_success_tackles(df,season,col):
    df1 = df[df.season==season]
    fig = go.Figure(data=[    
    #go.Bar(name='player_successful_raids', x=df1.name, y=df1.success_raids),
    go.Bar(name='team_name', x=df1.team_name, y=df1[col],
             text=(df1.name),textposition='auto')
    ])
    fig.update_layout(showlegend=True,title_text='Season : '+str(season) ,title=dict(x=0.5,y=0.9))
#     fig.update_layout(legend_orientation="h",legend=dict(x=0, y=-0.2))
    fig.update_layout(barmode='stack',xaxis={'categoryorder':'array', 'categoryarray':['Puneri Paltan','Jaipur Pink Panthers','Gujarat Fortunegiants']})
    fig.show()

In [0]:
def team_player_plot_spt(df,season,col):
    df1 = df[df.season==season]
    fig = go.Figure(data=[    
    #go.Bar(name='player_successful_raids', x=df1.name, y=df1.success_raids),
    go.Bar(name='team_name', x=df1.team_name, y=df1[col])
    ])
    fig.update_layout(showlegend=True,title_text='Season : '+str(season) ,title=dict(x=0.5,y=0.9))
#     fig.update_layout(legend_orientation="h",legend=dict(x=0, y=-0.2))
    fig.update_layout(barmode='stack',xaxis={'categoryorder':'array', 'categoryarray':['Puneri Paltan','Jaipur Pink Panthers','Gujarat Fortunegiants']})
    fig.show()

In [0]:
def team_player_plot_unsuccess_raids(df,season):
    df1 = df[df.season==season]
    #df1 = df1.sort_values(by=['team_successful_raids'],ascending=False)
    fig = go.Figure(data=[    
    #go.Bar(name='player_successful_raids', x=df1.name, y=df1.success_raids),
    go.Bar(name='team_name', x=df1.team_name, y=df1.unsuccess_raids,
             text=(df1.name),textposition='auto')
    ])
    fig.update_layout(showlegend=True,title_text='Season : '+str(season) ,title=dict(x=0.5,y=0.9))
#     fig.update_layout(legend_orientation="h",legend=dict(x=0, y=-0.2))
    fig.update_layout(barmode='stack',xaxis={'categoryorder':'array', 'categoryarray':['Dabang Delhi K.C.','Bengal Warriors','Bengaluru Bulls','Haryana Steelers']})
    fig.show()

In [0]:
def team_player_plot_super_raids(df,season):
    df1 = df[df.season==season]
    #df1 = df1.sort_values(by=['team_successful_raids'],ascending=False)
    fig = go.Figure(data=[    
    #go.Bar(name='player_successful_raids', x=df1.name, y=df1.success_raids),
    go.Bar(name='team_name', x=df1.team_name, y=df1.super_raids,
             text=(df1.name),textposition='auto')
    ])
    fig.update_layout(showlegend=True,title_text='Season : '+str(season) ,title=dict(x=0.5,y=0.9))
#     fig.update_layout(legend_orientation="h",legend=dict(x=0, y=-0.2))
    fig.update_layout(barmode='stack',xaxis={'categoryorder':'array', 'categoryarray':['Dabang Delhi K.C.','Bengal Warriors','Bengaluru Bulls','Haryana Steelers']})
    fig.show()

In [0]:
def team_player_plot_super_raids(df,season):
    df1 = df[df.season==season]
    #df1 = df1.sort_values(by=['team_successful_raids'],ascending=False)
    fig = go.Figure(data=[    
    #go.Bar(name='player_successful_raids', x=df1.name, y=df1.success_raids),
    go.Bar(name='team_name', x=df1.team_name, y=df1.super_raids,
             text=(df1.name),textposition='auto')
    ])
    fig.update_layout(showlegend=True,title_text='Season : '+str(season) ,title=dict(x=0.5,y=0.9))
#     fig.update_layout(legend_orientation="h",legend=dict(x=0, y=-0.2))
    fig.update_layout(barmode='stack',xaxis={'categoryorder':'array', 'categoryarray':['Dabang Delhi K.C.','Bengal Warriors','Bengaluru Bulls','Haryana Steelers']})
    fig.show()

In [29]:
player_all_data.head()

player_id  player_team  ... succ_raid_percent  succ_tackle_percent
0        105            0  ...         80.952381             7.692308
1        123            0  ...          0.000000            29.230769
2        156            0  ...         57.142857            14.285714
3         54            0  ...          0.000000            51.612903
4         71            0  ...        100.000000            33.707865

[5 rows x 42 columns]

In [35]:
player_plot_per_season(player_all_data,49)

In [37]:
player_plot_per_season(player_all_data,26)

In [39]:
team_raid_plot_per_season(team_all_data,49)

As we can see from above plot , that we have **Dabang Delhi K.C.** at the top of the table with **340 points** . **Bengal Warriors and Bengaluru Bulls** are not far behind and have **318 and 309 points** respectively. **We will select these teams and also Haryana Steelers for our analysis as they have also qualified for the semifinals .** 

In [40]:
team_raid_plot_per_season(team_all_data,26)

As we can see from above plot , **Bengaluru Bulls and Dabang Delhi K.C were in top 2 of successful raids and teams total points too.**

In [41]:
team_raid_plot_per_season(team_all_data,8)

**The 5 season was very skewed and favourable to Patna Pirates , thats why we cant take much ahead from this plot.**

**Our top teams for analysing Highest Successful Raid Points are** (based on current season) :

1.   Dabang Delhi K.C. (Q)
2.   Bengal Warriors (Q)
3.   Bengaluru Bulls
4.   Haryana Steelers (Q)



Now we will analyse the players contribution to the team successful raid points

In [0]:
list_of_teams = ['Dabang Delhi K.C.','Bengal Warriors','Bengaluru Bulls','Haryana Steelers']
player_all_data_filtered = player_all_data[player_all_data['team_name'].isin(list_of_teams)]
player_all_data_filtered = player_all_data_filtered[player_all_data_filtered['success_raids'] > 20]


In [44]:
team_player_plot_success_raids(player_all_data_filtered,49)

### Observations

*   We can clearly **see that number of contributors at a minimum of 20 points        capping are more or less same for all the teams into consideration**.
*   **Bengaluru Bulls have less chance** to score more because of **less contributors**
*   **Dabang Delhi** has the **single greater contributor Naveen Kumar with 201 points**, thats more     **than 50 % to team's total raid points** 
*   **Bengal Warriors has compartively more even contributions to the team's          score**, although Pawan has decently good record.






In [76]:
team_player_plot_success_raids(player_all_data_filtered,26)

In [77]:
team_player_plot_success_raids(player_all_data_filtered,8)

### Observations for Success Raids in Season 6

*   **Dabang Delhi K.C.** has **sparingly more contributors** with no spiking in contributions. The top player last year was **Naveen Kumar** for them.
* **Bengal Warriors** has spiking contribution from **Maninder Singh**.
*   **Bengaluru Bulls** has only two players who contributed more than 20 points to team's total raid score which is not a good indication. Pawan was the top player.





## Plotting Unsuccess Raids For Teams

In [47]:
team_player_plot_unsuccess_raids(player_all_data_filtered,49)

### Observations for unsuccessful raid for current season

*   **Dabang Delhi K.C. and Bengal Warriors **are neck to neck in terms of **unsuccessful raids** with contributions from all players who have high success raid points
*   **Bengaluru Bulls** have the **highest unsuccessful raid points currently which decreases their chances for a successful raid**.



In [48]:
team_player_plot_unsuccess_raids(player_all_data_filtered,26)

## Plotting for Super Raids

In [50]:
team_player_plot_super_raids(player_all_data_filtered,49)

### Observations

Bengal Warriors have higher super raids as compared to other teams.

In [52]:
team_player_plot_super_raids(player_all_data_filtered,26)

## Predicting the Team with Highest Successful Tackle Percentage

In [54]:
team_tackle_plot_per_season(team_all_data,49)

In [78]:
team_tackle_plot_per_season(team_all_data,26)

In [79]:
team_tackle_plot_per_season(team_all_data,8)

### Observations:

*   Teams are pretty close with tackle points **with Pune , Jaiupur and Gujarat**       in top three teams.
*   We will proceed with our analysis with the following teams :

  1. Puneri Paltan
  2. Jaipur Pink Panthers
  3. Gujarat Fortunegiants



In [0]:
list_of_teams_tackles = ['Puneri Paltan','Jaipur Pink Panthers','Gujarat Fortunegiants']
player_all_data_tackle_filtered = player_all_data[player_all_data['team_name'].isin(list_of_teams_tackles)]
player_all_data_tackle_filtered = player_all_data_tackle_filtered[player_all_data_tackle_filtered['success_tackles'] > 20]

In [57]:
team_player_plot_success_tackles(player_all_data_tackle_filtered,49,'success_tackles')

### Observations: 

*   **Puneri Paltan has more players to contribute** on the basis of filtering out players who have scored more than 20 tackle points
*   **Jaipur** **has the player (Sandeep Kumar Dhull) with the highest tackle points but its more skewed compared to other contributions.**



In [60]:
team_player_plot_success_tackles(player_all_data_tackle_filtered,26,'success_tackles')

In [81]:
team_player_plot_success_tackles(player_all_data_tackle_filtered,8,'success_tackles')

### Observations:

1. Gujarat was the clear winner in terms of tackle points , but the data looks to be skewed.
2. Puneri Paltan and Jaipur Pink Panthers were neck to neck last year .

In [75]:
team_player_plot_success_tackles(player_all_data_tackle_filtered,49,'success_tackles_per_match')

In [62]:
team_player_plot_success_tackles(player_all_data_tackle_filtered,26,'success_tackles_per_match')

In [63]:
team_player_plot_success_tackles(player_all_data_tackle_filtered,49,'super_tackles')

In [64]:
team_player_plot_success_tackles(player_all_data_tackle_filtered,26,'super_tackles')

In [65]:
team_player_plot_success_tackles(player_all_data_tackle_filtered,49,'unsuccesful_tackles')

In [66]:
team_player_plot_success_tackles(player_all_data_tackle_filtered,26,'unsuccesful_tackles')

In [67]:
team_player_plot_success_tackles(player_all_data_tackle_filtered,49,'succ_tackle_percent')

In [68]:
team_player_plot_success_tackles(player_all_data_tackle_filtered,26,'succ_tackle_percent')

## Prediction for Highest SPT

In [69]:
team_player_plot_spt(team_all_data,49,'spt')

In [82]:
team_player_plot_spt(team_all_data,8,'spt')

In [70]:
team_player_plot_spt(team_all_data,26,'spt')

In [71]:
team_player_plot_spt(team_all_data,49,'team_super_raid')

In [83]:
team_player_plot_spt(team_all_data,26,'team_super_raid')

In [84]:
team_player_plot_spt(team_all_data,8,'team_super_raid')

In [72]:
team_player_plot_spt(team_all_data,49,'team_super_tackles')

In [85]:
team_player_plot_spt(team_all_data,26,'team_super_tackles')

In [86]:
team_player_plot_spt(team_all_data,8,'team_super_tackles')

In [73]:
team_player_plot_spt(team_all_data,49,'team_all_outs_inflicted')

In [87]:
team_player_plot_spt(team_all_data,26,'team_all_outs_inflicted')

In [88]:
team_player_plot_spt(team_all_data,8,'team_all_outs_inflicted')

In [74]:
team_player_plot_spt(team_all_data,49,'team_all_outs_conceded')

In [89]:
team_player_plot_spt(team_all_data,26,'team_all_outs_conceded')

In [90]:
team_player_plot_spt(team_all_data,8,'team_all_outs_conceded')